onnx/test/shape_inference_test.py

```diff
@@ -293,6 +293,126 @@ def test_cast_like(self) -> None:
             graph, [make_tensor_value_info("y", TensorProto.FLOAT16, (2, 4, 3))]
         )
 
+    def test_col2im(self) -> None:
+        graph = self._make_graph(
+            [
+                ("input", TensorProto.FLOAT, (1, 5, 5)),
+                ("output_shape", TensorProto.INT64, (2,)),
+                ("kernel_shape", TensorProto.INT64, (2,)),
+            ],
+            [
+                make_node(
+                    "Col2Im", ["input", "output_shape", "kernel_shape"], ["output"]
+                )
+            ],
+            [],
+            initializer=[
+                make_tensor("output_shape", TensorProto.INT64, (2,), (5, 5)),
+                make_tensor("kernel_shape", TensorProto.INT64, (2,), (1, 5)),
+            ],
+        )
+        self._assert_inferred(
+            graph, [make_tensor_value_info("output", TensorProto.FLOAT, (1, 1, 5, 5))]
+        )
+
+    def test_col2im_strides(self) -> None:
+        graph = self._make_graph(
+            [
+                ("input", TensorProto.FLOAT, (1, 9, 4)),
+                ("output_shape", TensorProto.INT64, (2,)),
+                ("kernel_shape", TensorProto.INT64, (2,)),
+            ],
+            [
+                make_node(
+                    "Col2Im",
+                    ["input", "output_shape", "kernel_shape"],
+                    ["output"],
+                    strides=[2, 2],
+                )
+            ],
+            [],
+            initializer=[
+                make_tensor("output_shape", TensorProto.INT64, (2,), (5, 5)),
+                make_tensor("kernel_shape", TensorProto.INT64, (2,), (3, 3)),
+            ],
+        )
+        self._assert_inferred(
+            graph, [make_tensor_value_info("output", TensorProto.FLOAT, (1, 1, 5, 5))]
+        )
+
+    def test_col2im_pads(self) -> None:
+        graph = self._make_graph(
+            [
+                ("input", TensorProto.FLOAT, (1, 5, 15)),
+                ("output_shape", TensorProto.INT64, (2,)),
+                ("kernel_shape", TensorProto.INT64, (2,)),
+            ],
+            [
+                make_node(
+                    "Col2Im",
+                    ["input", "output_shape", "kernel_shape"],
+                    ["output"],
+                    pads=[0, 1, 0, 1],
+                )
+            ],
+            [],
+            initializer=[
+                make_tensor("output_shape", TensorProto.INT64, (2,), (5, 5)),
+                make_tensor("kernel_shape", TensorProto.INT64, (2,), (1, 5)),
+            ],
+        )
+        self._assert_inferred(
+            graph, [make_tensor_value_info("output", TensorProto.FLOAT, (1, 1, 5, 5))]
+        )
+
+    def test_col2im_dilations(self) -> None:
+        graph = self._make_graph(
+            [
+                ("input", TensorProto.FLOAT, (1, 4, 5)),
```

Comment:
Not sure how you get this shape for the given dilation?

Reply:
````python
import torch

kernel_size = (2, 2)
dilation = (1, 5)
output_size = (6, 6)

image = torch.arange(1, 37).reshape(1, 1, 6, 6).to(torch.float32)
print(f'User Image:\n{image}\n\t with shape {image.shape}')

# Im2Col used to generate input for target Col2Im
im2col = torch.nn.Unfold(kernel_size=kernel_size,
                         dilation=dilation)

col = im2col(image)
print(f'Im2Col output:\n{col}\n\t with shape {col.shape}')

# Col2Im results
col2im = torch.nn.Fold(kernel_size=kernel_size,
                       dilation=dilation,
                       output_size=output_size)
im = col2im(col)
print(f'Col2Im output:\n{im}\n\t with shape {im.shape}')
```

which returns

```python
User Image:
tensor([[[[ 1.,  2.,  3.,  4.,  5.,  6.],
          [ 7.,  8.,  9., 10., 11., 12.],
          [13., 14., 15., 16., 17., 18.],
          [19., 20., 21., 22., 23., 24.],
          [25., 26., 27., 28., 29., 30.],
          [31., 32., 33., 34., 35., 36.]]]])
         with shape torch.Size([1, 1, 6, 6])
Im2Col output:
tensor([[[ 1.,  7., 13., 19., 25.],
         [ 6., 12., 18., 24., 30.],
         [ 7., 13., 19., 25., 31.],
         [12., 18., 24., 30., 36.]]])
         with shape torch.Size([1, 4, 5])
Col2Im output:
tensor([[[[ 1.,  0.,  0.,  0.,  0.,  6.],
          [14.,  0.,  0.,  0.,  0., 24.],
          [26.,  0.,  0.,  0.,  0., 36.],
          [38.,  0.,  0.,  0.,  0., 48.],
          [50.,  0.,  0.,  0.,  0., 60.],
          [31.,  0.,  0.,  0.,  0., 36.]]]])
         with shape torch.Size([1, 1, 6, 6])
```

Reply:
Ok, thanks for the explanation! LGTM!